subprojects/docs/src/docs/userguide/upgrade_version_4.adoc

```diff
@@ -0,0 +1,493 @@
+ * The bundled `bndlib` has been upgraded from 3.2.0 to 3.4.0.
+ * The FindBugs Plugin no longer renders progress information from its analysis. If you rely on that output in any way, you can enable it with link:{groovyDslPath}/org.gradle.api.plugins.quality.FindBugs.html#org.gradle.api.plugins.quality.FindBugs:showProgress[FindBugs.showProgress].
+
+[[upgrade_4.0]]
+== From 4.0
+
+ * Consider using the new <<custom_tasks.adoc#worker_api,Worker API>> to enable units of work within your build to run in parallel.
+
+=== Deprecated classes, methods and properties
+
+Follow the API links to learn how to deal with these deprecations (if no extra information is provided here):
+
+ * link:{javadocPath}/org/gradle/api/tasks/scala/ScalaDocOptions.html#getStyleSheet--[ScalaDocOptions.styleSheet] — the Scaladoc Ant task in Scala 2.11.8 and later no longer support this property.
+ * link:{javadocPath}/org/gradle/api/Nullable.html[Nullable]
+ * link:{javadocPath}/org/gradle/api/Task.html#dependsOnTaskDidWork--[Task.dependsOnTaskDidWork()]
+
+=== Potential breaking changes
+
+ * Non-Java projects that have a <<dependency_types.adoc#sub:project_dependencies,project dependency>> on a Java project now consume the `runtimeElements` configuration by default instead of the `default` configuration.
++
+To override this behavior, you can explicitly declare the configuration to use in the project dependency. For example: `project(path: ':myJavaProject', configuration: 'default')`.
+ * Default Zinc compiler upgraded from 0.3.13 to 0.3.15.
+ * [Kotlin DSL] Base package renamed from `org.gradle.script.lang.kotlin` to `org.gradle.kotlin.dsl`.
+
+
+== Changes in detail
+
+[[rel4.8:lazy_task_creation]]
+=== [4.8] Consider trying the lazy API for task creation and configuration
+
+Gradle 4.9 introduced a new way to create and configure tasks that works lazily. When you use this approach for tasks that are expensive to configure, or when you have many, many tasks, your build configuration time can drop significantly when those tasks don't run.
+
+You can learn more about lazily creating tasks in the <<task_configuration_avoidance#task_configuration_avoidance,Task Configuration Avoidance>> chapter. You can also read about the background to this new feature in https://blog.gradle.org/preview-avoiding-task-configuration-time[this blog post].
+
+[[rel4.7:switch_to_publishing_plugins]]
+=== [4.7] Switch to the Maven Publish and Ivy Publish Plugins
+
+Now that the publishing plugins are stable, the <<artifact_management.adoc#artifact_management,legacy publishing>> mechanism based on `Upload` tasks is deprecated. You should replace `upload<Conf>` configuration with a `publishing` block instead. See the <<publishing_overview.adoc#publishing_overview,publishing overview chapter>> for information on how to use the publishing plugins.
```

Comment:
Why is it not deprecated? I hate to see the perpetuation of two publishing mechanisms.

Reply:
Because you can't define your own `Components` and so ecosystems like Android can't use the new publishing mechanism yet. They still rely on `Upload` until we provide that feature.

Reply:
Ah, yes, I remember about the custom components thing now. Thanks for the reminder. I hope that's implemented in the not too distant future!